kernel/base/src/main/java/com/twosigma/beakerx/kernel/magic/command/CodeFactory.java

```diff
@@ -26,102 +28,72 @@
-    Optional<MagicCommandFunctionality> mcOption = findMagicCommandFunctionality(kernel.getMagicCommandTypes(), firstCommand);
-    if (mcOption.isPresent()) {
-      magicCommands.add(new MagicCommand(mcOption.get(), firstCommand, restOfTheCode.toString()));
-    } else {
-      errors.add(processIllegalCommand("Cell magic " + firstCommand + " not found"));
+  private static CodeFrame createFrameForPlainCode(String line, Scanner scanner) {
+    List<String> result = new ArrayList<>();
+    result.add(line);
+    Pattern p = Pattern.compile("^%.*", Pattern.MULTILINE);
+    while (scanner.hasNext() && !scanner.hasNext(p)) {
+      String str = removeExtraWhitespaces(scanner.nextLine().trim());
+      result.add(str);
     }
-    return Code.createCodeWithoutCodeBlock(allCode, magicCommands, errors, message);
+    return new PlainCode(String.join(System.lineSeparator(), result));
   }
 
-  private static boolean isCellCommand(String command) {
-    return command.startsWith(CELL_COMMAND_MAGIC);
+  private static CodeFrame createFrameForLineMagic(String line, KernelFunctionality kernel) {
+    Optional<MagicCommandFunctionality> mcOption = findMagicCommandFunctionality(kernel.getMagicCommandTypes(), line);
+    return mcOption
+            .<CodeFrame>map(magicCommandFunctionality -> new MagicCommand(magicCommandFunctionality, line))
+            .orElseGet(() -> new ErrorCodeFrame(processIllegalCommand("Inline magic " + line + " not found")));
   }
 
-  private static Optional<String> codeWithoutCommands(Scanner scanner) {
-    List<String> codeWithoutCommands = new ArrayList<>();
-    while (scanner.hasNext()) {
-      codeWithoutCommands.add(scanner.nextLine());
-    }
-    if (codeWithoutCommands.isEmpty()) {
-      return Optional.empty();
-    }
-    return Optional.of(join(codeWithoutCommands, lineSeparator()));
+  private static CodeFrame createFrameForCellMagic(String line, Scanner scanner, KernelFunctionality kernel) {
+    Optional<MagicCommandFunctionality> mcOption = findMagicCommandFunctionality(kernel.getMagicCommandTypes(), line);
+    return mcOption
+            .<CodeFrame>map(magicCommandFunctionality -> new MagicCommand(magicCommandFunctionality, line, takeRestOfTheCode(scanner)))
+            .orElseGet(() -> new ErrorCodeFrame(processIllegalCommand("Cell magic " + line + " not found")));
   }
 
-  private static List<String> takeCommands(Scanner scanner) {
+  private static String takeRestOfTheCode(Scanner scanner) {
     List<String> result = new ArrayList<>();
-    Pattern p = Pattern.compile("^%.*", Pattern.MULTILINE);
-    while (scanner.hasNext(p)) {
-      result.add(removeExtraWhitespaces(scanner.nextLine()));
+    while (scanner.hasNextLine()) {
+      result.add(removeExtraWhitespaces(scanner.nextLine().trim()));
     }
-    return result;
+    return String.join(System.lineSeparator(), result);
   }
 
   private static String removeExtraWhitespaces(String command) {
```

Comment:
this can cause a problem if there are multiple spaces that actually mean something:
```
%classpath add jar "path/with  two_consecutive_spaces.jar"
```
also breaks
```
%time println("x  y")
```